x/programs/cmd/simulator/src/simulator.rs

```diff
@@ -0,0 +1,238 @@
+
+        self.responses
+            .next()
+            .ok_or(SimulatorError::Client(ClientError::Eof))?
+    }
+}
+
+/// A [`SimulatorRequest`] is a call to the simulator
+#[derive(Debug, Serialize, PartialEq, Clone)]
+#[serde(rename_all = "camelCase")]
+struct SimulatorRequest {
+    /// The API endpoint to call.
+    endpoint: Endpoint,
+    /// The method to call on the endpoint.
+    method: String,
+    /// The maximum number of units the Request can consume.
+    max_units: u64,
+    /// The parameters to pass to the method.
+    params: Vec<Param>,
+}
+
+impl SimulatorRequest {
+    pub fn read(method: String, params: Vec<Param>) -> Self {
+        Self {
+            endpoint: Endpoint::ReadOnly,
+            method: method.to_string(),
+            max_units: 0,
+            params,
+        }
+    }
+
+    pub fn execute(method: String, params: Vec<Param>, max_units: u64) -> Self {
+        Self {
+            endpoint: Endpoint::Execute,
+            method,
+            max_units,
+            params,
+        }
+    }
+
+    pub fn create_program(path: String) -> Self {
+        Self {
+            endpoint: Endpoint::CreateProgram,
+            // TODO: this does not need to be passed in
+            method: "create_program".to_string(),
+            max_units: 0,
+            params: vec![Param::Path(path)],
+        }
+    }
+}
+
+type SimulatorResponseItem = Result<SimulatorResponse, SimulatorError>;
+
+#[derive(Debug, Deserialize)]
+pub struct SimulatorResult {
+    /// The ID created from the program execution.
+    pub action_id: Option<String>,
+    /// The timestamp of the function call response.
+    pub timestamp: u64,
+    /// The result of the function call.
+    #[serde(deserialize_with = "base64_decode")]
+    response: Vec<u8>,
+}
+
+impl SimulatorResult {
+    pub fn response<T>(&self) -> Result<T, SimulatorResponseError>
+    where
+        T: BorshDeserialize,
+    {
+        let res: Result<T, ExternalCallError> = borsh::from_slice(&self.response)?;
+        res.map_err(SimulatorResponseError::ExternalCall)
+    }
+}
+
+#[derive(Debug, Deserialize)]
+pub struct SimulatorResponse {
+    /// The numeric id of the response.
+    #[serde(deserialize_with = "id_from_usize")]
+    pub id: Id, // TODO override of the Id Deserialize before removing the prefix
+    /// An optional error message.
+    pub error: Option<String>,
+    pub result: SimulatorResult,
+}
+
+/// The endpoint to call for a [`SimulatorRequest`].
+#[derive(Debug, Serialize, PartialEq, Clone)]
+#[serde(rename_all = "lowercase")]
+enum Endpoint {
+    /// Make a read-only call to a program function and return the result.
+    ReadOnly,
+    /// Create a transaction on-chain from a possible state changing program
+    /// function call. A program's function can internally optionally call other
+    /// functions including program to program.
+    Execute,
+    /// Create a new program on-chain
+    CreateProgram,
+}
+
+use std::path::Path;
```

Comment:
Please import this at the top of the file

Reply:
moved it down b/c @richardpringle comment
https://github.com/ava-labs/hypersdk/pull/1166#discussion_r1683354975

Reply:
I see, you might as well scope it inside of the `get_binary_path` function since it's only used here.